keps/sig-network/4631-loadbalancerstatus/README.md

```diff
@@ -0,0 +1,1210 @@
+way, the cloud provider must update at least the `LastTransitionTime`
+of the `LoadBalancerServing` condition. (For example, if the user
+changes the service's `LoadBalancerSourceRanges`, and the cloud
+provider updates the load balancer for this, it must update the
+`LoadBalancerServing` condition's `LastTransitionTime`, even if the
+value of `LoadBalancerServing` did not toggle from `True` to `False`
+and back to `True`.)
+
+#### The `LoadBalancerProvisioning` Condition
+
+The new `LoadBalancerProvisioning` condition indicates that the cloud
+provider is currently performing a time-consuming
+resource-provisioning step for the load balancer. When
+`LoadBalancerProvisioning` is `True`, the load balancer is either not
+handling connections, or else is still handling connections according
+to the semantics of a previous state of the Service (with the value of
+the `LoadBalancerServing` condition distinguishing these two cases).
+
+The cloud provider should set a `LoadBalancerProvisioning` condition
+immediately upon first becoming aware of a `Type: LoadBalancer`
+Service. At all times, the condition should be `True` if the cloud
+provider is waiting for the completion of an operation to bring the
+load balancer's behavior in line with the current state of the Service
+object, and `False` if it not. (There are no defined semantics for an
+`Unknown` value of this condition.)
+
+Any time any load-balancer-related field (or provider-specific
+annotation) of the Service object changes, the cloud provider must
+update at least the `LastTransitionTime` of the
+`LoadBalancerProvisioning` condition, to indicate that it has seen the
+update, even if the value of `LoadBalancerProvisioning` remains
+`False`.
+
+#### Terminating Condition
+
+```
+<<[UNRESOLVED termination ]>>
+
+Do we need/want a `LoadBalancerTerminating` condition? And/or, should
+the cloud provider set `LoadBalancerProvisioning` when it is
+*de*-provisioning a load balancer?
+
+There is possibly no need for this when deleting a Service, but it may
+help in the case of converting a Service from `Type: LoadBalancer` to
+`Type: ClusterIP` (which is tested by the e2es).
+
+<<[/UNRESOLVED]>>
+```
+
+#### Service Lifecycle and Interaction of the Conditions
+
+When a load balancer Service is first created (or a non-load-balancer
+Service is converted to `Type: LoadBalancer`), one of three things
+should immediately happen:
+
+  1. The cloud provider sets `LoadBalancerProvisioning=False`,
+     `LoadBalancerServing=False`, with appropriate `Reason` values
+     (discussed below), indicating that it cannot or will not
+     provision a load balancer for the service.
+
+  2. The cloud provider sets `LoadBalancerProvisioning=False`,
+     `LoadBalancerServing=True`, (and fills in `.Status.LoadBalancer`)
+     indicating that the load balancer for the service is already
+     available.
+
+  3. The cloud provider sets `LoadBalancerProvisioning=True`,
```

Comment:
What about `(true, true)` ?  Conditions are best when they are orthogonal.  These are not.  We have to spec it, since both fields COULD be set at the same time.

IOW, don't let this become a state-machine API.  It's a collection of roughly independent observations.

I see it more clearly below, so now this feels duplicative

Reply:
This is actually specified below; (true, true) means it is both serving and provisioning. ie, it is continuing to serve while reprovisioning for an update.

> IOW, don't let this become a state-machine API. It's a collection of roughly independent observations.

I feel like we definitely want the "provisioning" observation. I feel like the "serving" observation is also useful? Did you have some other idea for conditions?